### PR TITLE
phase 3 initial

### PR DESCRIPTION
**Pull request by Rompue**

### CODE REVIEW

**Comment by Rompue** [comment]
@barath @ralphchung

**Comment by barath** [comment]
Could you shorten this pull request?  (Ideal length is 150 diff lines.)  Let me know if not.

**Comment by Rompue** [comment]
@barath Could you only review half of the pull request this time, if limited to 150? It is kind of hard to revert the changes...

**Comment by Rompue** [comment]
> Could you shorten this pull request? (Ideal length is 150 diff lines.) Let me know if not.

I deleted some changes. Is it okay now?

File: src/service_client_lib.h [hunk 1]
```diff
@@ -89,6 +89,12 @@ class ServiceClient : public GrpcClient<chirp::ServiceLayer::Stub> {
   ReturnCodes SendMonitorRequest(const std::string &username,
                                  std::vector<Chirp> *const chirps);
 
+  // Send a stream request to the server with the tag `tag`
+  // returns OK if this operation succeeds
+  // returns other error codes if this operation fails
+  ReturnCodes SendStreamRequest(const std::string &tag,
+                                 std::vector<Chirp> *const chirps);
```
Comment (barath) [inline src/service_client_lib.h]: This signature is a bit funny; why a const pointer of vector?

File: src/service_data_structure.cc [hunk 2]
```diff
@@ -201,6 +201,28 @@ ServiceDataStructure::ReturnCodes ServiceDataStructure::UserSession::PostChirp(
     return INTERNAL_BACKEND_ERROR;
   }
 
+
+
+  // Parse the chirp text to find any tags
+  std::string::size_type start = text.find('#');
```
Comment (barath) [inline src/service_data_structure.cc]: I would recommend moving the string parsing to a helper.

File: src/service_data_structure.cc [hunk 3]
```diff
@@ -201,6 +201,28 @@ ServiceDataStructure::ReturnCodes ServiceDataStructure::UserSession::PostChirp(
...
+      // insert an entry
+      std::string::size_type count = end != std::string::npos ? end - start : text.size() - start;
+      UserChirpList chirp_tag_list;
+      ok = chirp_connect_backend::GetChirpTagList(text.substr(start, count),
+                                                   &chirp_tag_list);
+      chirp_tag_list.insert(chirp.get_id());
+      ok = chirp_connect_backend::SaveChirpTag(text.substr(start, count), chirp_tag_list);
+      start++;
+    }
+    if (end == std::string::npos) break;
```
Comment (barath) [inline src/service_data_structure.cc]: Braces around if body (break).

File: src/service_client_lib.h (hunk 1 above)
Comment (Rompue) [reply src/service_client_lib.h]: Fixed

File: src/service_data_structure.cc (hunk 2 above)
Comment (Rompue) [reply src/service_data_structure.cc]: Fixed.

File: src/service_data_structure.cc (hunk 3 above)
Comment (Rompue) [reply src/service_data_structure.cc]: Fixed.

File: src/service_client_lib.h (hunk 1 above)
Comment (Rompue) [reply src/service_client_lib.h]: Because the function shouldn't modify the content of the pointted vector. It is issued from another function.

Comment (Rompue) [reply src/service_client_lib.h]: Because the function modifies the content of the monitor but cannot point the vector to another one.

**Pull request merged**